### DISCUSSION TITLE
JDK19 - javax.sql.rowset.RowSetFactory

### DISCUSSION THREAD

**Discussion by jcurry5** [nextgenhealthcare/connect#5530] [General]
Using the previous connection pool MIrth implementation, it is no longer usable for JDK19 (I believe even starting from JDK16 but don't quote me on that).

Looking at the logs, it looks like the following has to be updated:
```
var db_res = statement.executeQuery();
valOut = new Packages.com.sun.rowset.CachedRowSetImpl();
valOut.populate(db_res);
statement.close();
```

The official documentation is saying to implement 			
```
RowSetFactory factory = RowSetProvider.newFactory();
CachedRowSet rowset = factory.createCachedRowSet();
```
But factory and rowset are not resolving, even after explicitly using import and importPackage.

Any insight on what I'm overlooking?

Thanks,

**Comment by jonbartels** [comment]
MC does run on at least Java 17. I have containers running on it literally right now. I recall running into an issue with Java 18 or 19 but I never bothered to figure it out because it wasn't an LTS Java version.

Does https://github.com/nextgenhealthcare/connect/issues/5367 from MC 4.1.1 resolve this?  The related commit is https://github.com/nextgenhealthcare/connect/commit/d3e31092af148d3c341ee4d2f902222019cc2bd5 . 

You said you found logs, can you show what log messages or errors you saw?

What MC version are you running?

Also did you follow the instructions for Java 9 and newer at https://docs.nextgen.com/bundle/Mirth_User_Guide_41/page/connect/connect/topics/c_Using_Java_9_or_greater_mirth_connect_ug.html ? I assume you did but I have to ask!

**Reply by jonbartels** [reply]
Wait - I could be a moron. You're talking about changing *channel* code to work in Java 19 not getting MC to run.

**Reply by jcurry5** [reply]
> Wait - I could be a moron. You're talking about changing _channel_ code to work in Java 19 not getting MC to run.

Yeah that's right.  MC works fine using JDK19 and runs fine but the ConnectionPool code that uses new Packages.com.sun.rowset.CachedRowSetImpl(); doesn't work.  And from the updated JavaDocs it should be using 			RowSetFactory factory = RowSetProvider.newFactory();, CachedRowSet rowset = factory.createCachedRowSet(); but factory and rowset don't resolve even after importing the packages in the CP function.

**Reply by jcurry5** [reply]
> MC does run on at least Java 17. I have containers running on it literally right now. I recall running into an issue with Java 18 or 19 but I never bothered to figure it out because it wasn't an LTS Java version.
> 
> Does #5367 from MC 4.1.1 resolve this? The related commit is [d3e3109](https://github.com/nextgenhealthcare/connect/commit/d3e31092af148d3c341ee4d2f902222019cc2bd5) .
> 
> You said you found logs, can you show what log messages or errors you saw?
> 
> What MC version are you running?
> 
> Also did you follow the instructions for Java 9 and newer at https://docs.nextgen.com/bundle/Mirth_User_Guide_41/page/connect/connect/topics/c_Using_Java_9_or_greater_mirth_connect_ug.html ? I assume you did but I have to ask!

I'm running 4.0 so let me try 4.1.1.

Here's the log:
```
valOut = new Packages.com.sun.rowset.CachedRowSetImpl()
	
	else throw "updatePoolDbConn(): dbConn.prepareStatement() failed";LINE NUMBER:
	175DETS:
	Wrapped java.lang.IllegalAccessException: class org.mozilla.javascript.MemberBox cannot access class com.sun.rowset.CachedRowSetImpl (in module java.sql.rowset) because module java.sql.rowset does not export com.sun.rowset to unnamed module @4ae1a766
```

**Comment by jcurry5** [comment]
Thanks, Jon.  That did the trick!

I had to updated to 4.1.1 and change:
`valOut = new Packages.com.sun.rowset.CachedRowSetImpl();`
to
`valOut = new MirthCachedRowSet();`
